### ISSUE TITLE
Investigate fork selection equivalence mismatch

### ISSUE THREAD

**Issue by adinapoli-mndc** [oscoin/oscoin#429]
From this [build](https://console.cloud.google.com/cloud-build/builds/c671143e-2055-4d53-b662-d392c22dc782?project=opensourcecoin): 

```
Storage.Block
SQLite Storage blackbox
Use --quickcheck-replay=917375 to reproduce.
Block {blockHeader = BlockHeader {blockPrevHash = "R1ELJ1v", blockDataHash = "Pji2uFs", blockStateHash = "4J2q5RT", blockTimestamp = Timestamp 42390614099162, blockTargetDifficulty = Difficulty {fromDifficulty = 0}, blockSeal = SealedWith "</"}, blockHash = "YkoZBET", blockData = fromList [Tx {txMessage = Signed {sigMessage = Annotated (WithPos (InternalPos "src/Radicle/Internal/Annotation.hs:79:28 thisPos\nsrc/Radicle/Internal/Annotation.hs:36:50 toAnnotation") (NumberF ((-1852296587708) % 3229532965117))), sigSignature = MockSignature [94,94,219,71,241,245,106,213]}, txPubKey = MockPK (PK (MockKey [94,94,219,71,241,245,106,213]) (Hashed "hFqXpv6")), txChainId = 1, txNonce = 2, txContext = "8mni7zR"},Tx {txMessage = Signed {sigMessage = Annotated (WithPos (InternalPos "src/Radicle/Internal/Annotation.hs:79:28 thisPos\nsrc/Radicle/Internal/Annotation.hs:36:50 toAnnotation") (NumberF (5222099247581 % 3418035084665))), sigSignature = MockSignature [29,14,55,93,211,139,73,7]}, txPubKey = MockPK (PK (MockKey [29,14,55,93,211,139,73,7]) (Hashed "JQ1cYrs")), txChainId = 4, txNonce = 2, txContext = "PeFdHk1"}]} /= Block {blockHeader = BlockHeader {blockPrevHash = "fiu6biF", blockDataHash = "CQZouvr", blockStateHash = "W3sdJFe", blockTimestamp = Timestamp 42831309040803, blockTargetDifficulty = Difficulty {fromDifficulty = 0}, blockSeal = SealedWith "\FS\727784;\ESCk\697266H\FSJ>a9"}, blockHash = "ZnqcE1e", blockData = fromList [Tx {txMessage = Signed {sigMessage = Annotated (WithPos (InternalPos "src/Radicle/Internal/Annotation.hs:79:28 thisPos\nsrc/Radicle/Internal/Annotation.hs:36:50 toAnnotation") (NumberF (1977492603782 % 288807972433))), sigSignature = MockSignature [73,38,184,8,190,126,136,60]}, txPubKey = MockPK (PK (MockKey [73,38,184,8,190,126,136,60]) (Hashed "iGEWgEr")), txChainId = 98, txNonce = 109, txContext = "Yk5tGf7"},Tx {txMessage = Signed {sigMessage = Annotated (WithPos (InternalPos "src/Radicle/Internal/Annotation.hs:79:28 thisPos\nsrc/Radicle/Internal/Annotation.hs:36:50 toAnnotation") (NumberF ((-28201519195159) % 3736384581802))), sigSignature = MockSignature [6,125,56,159,174,249,10,139]}, txPubKey = MockPK (PK (MockKey [6,125,56,159,174,249,10,139]) (Hashed "cLA7EJa")), txChainId = 111, txNonce = 6, txContext = "E1jHfyJ"},Tx {txMessage = Signed {sigMessage = Annotated (WithPos (InternalPos "src/Radicle/Internal/Annotation.hs:79:28 thisPos\nsrc/Radicle/Internal/Annotation.hs:36:50 toAnnotation") (NumberF ((-355532849589) % 42963497534))), sigSignature = MockSignature [128,212,135,150,69,28,113,252]}, txPubKey = MockPK (PK (MockKey [128,212,135,150,69,28,113,252]) (Hashed "17b3Ar4")), txChainId = 96, txNonce = 12, txContext = "KGXECx3"},Tx {txMessage = Signed {sigMessage = Annotated (WithPos (InternalPos "src/Radicle/Internal/Annotation.hs:79:28 thisPos\nsrc/Radicle/Internal/Annotation.hs:36:50 toAnnotation") (NumberF ((-25767441461675) % 4030487530446))), sigSignature = MockSignature [70,125,36,162,72,82,30,88]}, txPubKey = MockPK (PK (MockKey [70,125,36,162,72,82,30,88]) (Hashed "23k7oFi")), txChainId = 45, txNonce = 32, txContext = "GACiqgJ"},Tx {txMessage = Signed {sigMessage = Annotated (WithPos (InternalPos "src/Radicle/Internal/Annotation.hs:79:28 thisPos\nsrc/Radicle/Internal/Annotation.hs:36:50 toAnnotation") (NumberF (60276902944458 % 6914840929649))), sigSignature = MockSignature [102,170,127,84,101,236,157,183]}, txPubKey = MockPK (PK (MockKey [102,170,127,84,101,236,157,183]) (Hashed "bnGW4UW")), txChainId = 23, txNonce = 53, txContext = "5nEfX4c"},Tx {txMessage = Signed {sigMessage = Annotated (WithPos (InternalPos "src/Radicle/Internal/Annotation.hs:79:28 thisPos\nsrc/Radicle/Internal/Annotation.hs:36:50 toAnnotation") (NumberF (137286829419554 % 9981493660417))), sigSignature = MockSignature [87,249,150,172,214,33,123,39]}, txPubKey = MockPK (PK (MockKey [87,249,150,172,214,33,123,39]) (Hashed "18Uj25r")), txChainId = 2, txNonce = 56, txContext = "LatecBL"},Tx {txMessage = Signed {sigMessage = Annotated (WithPos (InternalPos "src/Radicle/Internal/Annotation.hs:79:28 thisPos\nsrc/Radicle/Internal/Annotation.hs:36:50 toAnnotation") (BooleanF True)), sigSignature = MockSignature [239,85,164,55,35,110,253,211]}, txPubKey = MockPK (PK (MockKey [239,85,164,55,35,110,253,211]) (Hashed "MFP3qjd")), txChainId = 39, txNonce = 37, txContext = "W7Q1Kbi"}]}

Counterexample is:

pureStore = [ZnqcE1e (p:fiu6biF) @11h]

sqlStore = [YkoZBET (p:R1ELJ1v) @11h]

api call at test/Oscoin/Test/Storage/Block/Equivalence.hs:90:23 yielded a result mismatch!

- [YhzpbCv (p:18auNLF) @1h] link: [18auNLF (p:aCrGXLe) @52m]
- [CFAoYmu (p:fCLrkMk) @5h] ← [7JSLdiD (p:CFAoYmu) @6h] ← [c3doitG (p:7JSLdiD) @7h] ← [M7A3wDz (p:c3doitG) @8h] ← [TQu7CWu (p:M7A3wDz) @9h] ← [hwm9btQ (p:TQu7CWu) @10h] ← [cuq2wJM (p:hwm9btQ) @11h] ← [9HuczZR (p:cuq2wJM) @11h] link: [fCLrkMk (p:9baqMPC) @5h]
- [R1ELJ1v (p:Ek5hLc3) @10h] ← [YkoZBET (p:R1ELJ1v) @11h] link: [Ek5hLc3 (p:CRdNfUZ) @10h]
orphans:
chain: [aCrGXLe (p:1111111) @0ns] ← [fXrENhp (p:aCrGXLe) @52m] ← [iGdkJhJ (p:fXrENhp) @1h] ← [8BnLM8q (p:iGdkJhJ) @2h] ← [Gev6P2z (p:8BnLM8q) @3h] ← [9baqMPC (p:Gev6P2z) @4h] ← [dbzPAag (p:9baqMPC) @5h] ← [cMEU73D (p:dbzPAag) @5h] ← [5o4ZcM9 (p:cMEU73D) @6h] ← [V3NTieW (p:5o4ZcM9) @7h] ← [5bfvLRu (p:V3NTieW) @8h] ← [CRdNfUZ (p:5bfvLRu) @9h] ← [NNAKYgj (p:CRdNfUZ) @10h] ← [fiu6biF (p:NNAKYgj) @11h] ← [ZnqcE1e (p:fiu6biF) @11h]
*** Failed! Falsifiable (after 12 tests):
(forks) getTip . insertBlock equivalence: FAIL (3.54s)
```

**Comment by cloudhead** [comment]
I love how with quickcheck you never know when you might discover an error :joy:

**Comment by adinapoli-mndc** [comment]
I know right! In this particular example, it seems like the `Protocol` algorithm decided a new fork was found (tip: `[YkoZBET (p:R1ELJ1v) @11h]`) but the pure store disagrees, as the original chain stays unchanged on its end.

Who is right? 😅  (_me digs deeper_)

**Comment by adinapoli-mndc** [comment]
Ok, I have found the (subtle) culprit, and is yet another one introduced to punish us by our lack of `Oscoin.Time.Chrono` (but hey, now we have that 😅 ). Let's consider the offending chains, alongside their scores:

```
"FORK: [YkoZBET (p:R1ELJ1v) @11h] \8592 [R1ELJ1v (p:Ek5hLc3) @10h] \8592 [Ek5hLc3 (p:CRdNfUZ) @10h] SCORE: [0,0,0]
0"
"SUFFIX: [ZnqcE1e (p:fiu6biF) @11h] \8592 [fiu6biF (p:NNAKYgj) @11h] \8592 [NNAKYgj (p:CRdNfUZ) @10h] SCORE: [0,0,0]
0"
```

Here we see that we have a candidate which tip is `YkoZBET` and is compared against the adopted suffix with tip `ZnqcE1e`. Due to the fact both chains have same score (0) and same length (3), the system falls back to comparing the hash of their tip, and here is the screw up. Let's print out what the orphanage _thinks_ the tips are:

```
Just "YkoZBET"
Just "NNAKYgj"
```

What, nonsense! The culprit is that the `fromChainSuffix` is not typed enough:

```haskell
fromChainSuffix :: (Block c tx (Sealed c s) -> Score)
                -> NonEmpty (Block c tx (Sealed c s))
                -> ChainCandidate c s
fromChainSuffix scoreBlock blks = ChainCandidate
    { candidateChain = Seq.fromList . NonEmpty.toList . map blockHash $ blks
    , candidateTipHeader = blockHeader . NonEmpty.head $ blks
    , candidateScore = sum (map scoreBlock blks)
    }
```

Here we _don't know_ the order of blocks, and sure enough I have never noticed that inside the `stepProtocol` (and this is the algorithm which was extracted from `SQLite.Internal` and that ran for weeks) I feed into `fromChainSuffix` the chain fragment taken from the blockstore, which is ordered `NewestFirst`! Sure enough, the fix is simple: we need to make explicit things in the types:

```haskell
fromChainSuffix :: (Block c tx (Sealed c s) -> Score)
                -> OldestFirst NonEmpty (Block c tx (Sealed c s))
                -> ChainCandidate c s
fromChainSuffix scoreBlock (toOldestFirst -> blks) = ChainCandidate
    { candidateChain = Seq.fromList . NonEmpty.toList . map blockHash $ blks
    , candidateTipHeader = blockHeader . NonEmpty.head $ blks
    , candidateScore = sum (map scoreBlock blks)
    }
```

And pass the chain in the right order in `stepProtocol`:

```haskell
                 || bestFork > O.fromChainSuffix scoreFn (OldestFirst . NonEmpty.reverse . NonEmpty.fromList . toNewestFirst $ chainSuffix)
```

..and things start working again:

<img width="583" alt="Screenshot 2019-03-22 at 16 23 02" src="https://user-images.githubusercontent.com/45846748/54834007-8415da80-4cbf-11e9-8835-405540589ff1.png">


PR soon!

TL;DR `Oscoin.Time.Chrono` is very handy, so is QuickCheck!
